core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -556,30 +559,37 @@ public RequestLogBuilder logBuilder() {
 
     @Override
     public String toString() {
-        final String clientRequestIdStr = toStringClientRequestId();
-        if (strVal != null) {
-            return clientRequestIdStr + strVal;
+        if (strVal == null) {
+            strVal = toStringSlow();
+            if (channel() != null) {
+                strValAvailabilities |= STR_CHANNEL_AVAILABILITY;
+            }
+            if (log().parent() != null) {
+                strValAvailabilities |= STR_PARENT_LOG_AVAILABILITY;
+            }
```

Comment:
How about simplifying like this:

```java
final Channel ch = channel();
final RequestLogAccess parent = log().parent();
final short newAvailability = (ch != null ? ... : 0) | (parent != null ? ... : 0);
if (strVal != null && strValAvailabilities == newAvailability) {
    return strVal;
}

strValAvailabilities = newAvailability;
return toStringSlow(ch, parent);
```

Keep in mind that `toStringSlow()` does this:

```java
        if (ch != null) {
            this.strVal = strVal;
        }
```

.. and you must change it to make it work properly. Add tests properly so that this kind of mistake doesn't slip in.

Reply:
Thanks for the gentle suggestion. I've learned a lot. 🙏